### PR TITLE
[Android][TASKS][57661 + 57662] Read book UI, Load Book

### PR DESCRIPTION
**Pull request by ghost**
## Related Tickets
- [#Task 57661: Read book UI](https://edu-redmine.sun-asterisk.vn/issues/57661)
- [#Task 57662: Load book to UI](https://edu-redmine.sun-asterisk.vn/issues/57662)

## WHAT
- Parse data from epub file without library
- Custom a web view to support book-liked paging
- Display book on custom webview
- UI for book customization
## Review Checklist
https://youtube.com/shorts/kkzNJQgMLaI?feature=share
Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -41,4 +45,4 @@
             android:screenOrientation="portrait"></activity>
     </application>
 
-</manifest>
+</manifest>
```
Comment (minhnn-1279) [inline app/src/main/AndroidManifest.xml]: endline nhé a

Comment (hunggld) [inline app/src/main/AndroidManifest.xml:48]: endline nè a

File: app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/local/database/book/BookDbDAO.kt [hunk 2]
```diff
@@ -1,7 +1,9 @@
 package vn.ztech.software.projectgutenberg.data.repository.source.local.database.book
 
+import android.content.ContentValues
 import android.database.Cursor
 import android.provider.BaseColumns
+import android.util.Log
```
Comment (minhnn-1279) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/local/database/book/BookDbDAO.kt]: remove log a nhé :v

File: app/src/main/res/layout/item_toc.xml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        android:layout_margin="@dimen/dp_5"
+        android:textColor="?attr/colorOnPrimary"
+        app:layout_constraintStart_toStartOf="@id/ivDoneChecked"
+        app:layout_constraintEnd_toEndOf="@id/ivDoneChecked"
+        app:layout_constraintTop_toTopOf="@id/ivDoneChecked"
+        app:layout_constraintBottom_toBottomOf="@id/ivDoneChecked"
+        >
+    </TextView>
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (hungld-1302) [inline app/src/main/res/layout/item_toc.xml]: .

File: app/src/main/res/layout/layout_color_theme_settings.xml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (hungld-1302) [inline app/src/main/res/layout/layout_color_theme_settings.xml]: .

File: app/src/main/res/layout/layout_settings.xml [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            app:layout_constraintVertical_bias="0.5"
+            app:layout_constraintStart_toEndOf="@id/bt_cancel"
+            app:layout_constraintStart_toStartOf="parent">
+
+        </Button>
+
+    </androidx.constraintlayout.widget.ConstraintLayout>
+
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (hungld-1302) [inline app/src/main/res/layout/layout_settings.xml:100]: endline. Anh đặt lại id cho đúng convention nha

File: app/src/main/res/values/ids.xml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <item name="web_view_read_book" type="id" />
+</resources>
```
Comment (hungld-1302) [inline app/src/main/res/values/ids.xml]: .

File: app/src/main/java/vn/ztech/software/projectgutenberg/data/model/epub/NavPoint.kt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package vn.ztech.software.projectgutenberg.data.model.epub
+
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class NavPoint(
+    val playOrder: Int,
+    var navLabel: String = "", var content: String = ""
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/model/epub/NavPoint.kt]:
```suggestion
    var navLabel: String = "",
     var content: String = ""
```

File: app/src/main/java/vn/ztech/software/projectgutenberg/data/model/epub/NavPoint.kt [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+) : Parcelable {
+
+    constructor(playOrder: String) : this(Integer.parseInt(playOrder), "", "")
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/model/epub/NavPoint.kt]:
```suggestion
    constructor(playOrder: String) : this(playOrder.toInt(), "", "")
```

Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/model/epub/NavPoint.kt]:
```suggestion
    constructor(playOrder: String) : this(playOrder.toInt(), "", "")
```

File: app/src/main/java/vn/ztech/software/projectgutenberg/data/model/epub/TableOfContents.kt [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    fun clear() {
+        navPoints.clear()
+    }
+
+    fun getNavPoint(index: Int): NavPoint {
+        return navPoints[index]
+    }
+
+    fun getLatestPoint(): NavPoint {
+        return navPoints[navPoints.size - 1]
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/model/epub/TableOfContents.kt]:
```suggestion
        return navPoints.lastOrNull()
```

File: app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/local/BookLocalDataSource.kt [hunk 10]
```diff
@@ -3,15 +3,23 @@ package vn.ztech.software.projectgutenberg.data.repository.source.local
...
+import vn.ztech.software.projectgutenberg.data.repository.source.local.database.bookreading.BookReadingDbDAO
 import vn.ztech.software.projectgutenberg.data.repository.source.local.utils.AsynchronousCall
 import vn.ztech.software.projectgutenberg.data.repository.source.repository.book.BookDataSource
+import vn.ztech.software.projectgutenberg.utils.extension.getUnzipAbsolutePath
+import vn.ztech.software.projectgutenberg.utils.file.FileUtils
 
 
-class BookLocalDataSource(private val bookDbDAO: BookDbDAO) : BookDataSource.Local {
+class BookLocalDataSource(
+    private val bookDbDAO: BookDbDAO, private val bookReadingDbDAO: BookReadingDbDAO
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/local/BookLocalDataSource.kt]:
```suggestion
    private val bookDbDAO: BookDbDAO, 
    private val bookReadingDbDAO: BookReadingDbDAO
```

File: app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/local/BookLocalDataSource.kt [hunk 11]
```diff
@@ -74,12 +82,69 @@ class BookLocalDataSource(private val bookDbDAO: BookDbDAO) : BookDataSource.Loc
...
+        onResultListener: OnResultListener<EpubFile>
+    ) {
+        bookReadingDbDAO.writableDAO.storeToc(book, epubFile)
+    }
+
+    override fun getToc(bookId: Int, onResultListener: OnResultListener<Toc>) {
+        AsynchronousCall(
+            {
+                val toc = bookReadingDbDAO.readableDAO.getToc(bookId)
+                toc
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/local/BookLocalDataSource.kt]:
```suggestion
                bookReadingDbDAO.readableDAO.getToc(bookId)
```

File: app/src/main/java/vn/ztech/software/projectgutenberg/screen/download/DownloadFragment.kt [hunk 12]
```diff
@@ -70,10 +71,10 @@ class DownloadFragment
             var msg = ""
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/screen/download/DownloadFragment.kt]:
```suggestion
  val message = context?.getString(if(data) R.string.msg_delete_book_success else R.string.msg_delete_book_success )
```

File: app/src/main/java/vn/ztech/software/projectgutenberg/screen/download/DownloadFragment.kt [hunk 13]
```diff
@@ -84,6 +85,34 @@ class DownloadFragment
...
+
+        override fun onUnzipBookFailed(book: BookLocal) {
+            listBookLocalAdapter.setLoading(book, isLoading = false)
+        }
+
+        override fun onParseEpubDone(book: BookLocal, result: Result) {
+            when (result) {
+                Result.SUCCESS -> {
+                    Log.d("PARSEEPUB", "$result " + book.toString())
+                    context?.toast("Prepare book done! You can read the book now :D")
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/screen/download/DownloadFragment.kt]: remove log and extract string a ơi

File: app/src/main/java/vn/ztech/software/projectgutenberg/screen/readbook/ReadBookActivity.kt [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+                    Constant.LoadingAreaReadBook.ReadBookMain -> {
+                        binding?.layoutLoading?.layoutLoading?.visibility = visibility
+                    }
+                }
+            }
+        }
+
+    }
+
+    companion object {
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/screen/readbook/ReadBookActivity.kt]: companion object để cuối file a

File: app/src/main/java/vn/ztech/software/projectgutenberg/screen/readbook/WebViewHorizontal.kt [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+                currentToc?.let { listener?.onLoadChapterDone(it) }
+            }
+        }
+        val settings = settings
+        settings.cacheMode = WebSettings.LOAD_NO_CACHE
+        settings.builtInZoomControls = false
+        settings.allowFileAccess = true
+        settings.allowContentAccess = true
+        settings.displayZoomControls = false
+        settings.javaScriptEnabled = true
```
Comment (hungvv-1301) [inline app/src/main/java/vn/ztech/software/projectgutenberg/screen/readbook/WebViewHorizontal.kt]: scope func

File: app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/local/database/bookreading/BookReadingDbDAOInterface.kt [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+//    ): Boolean
+//
+//    fun removeBook(id: Int): Boolean
+//    fun clearNotExistItems(books: List<BookLocal>): Boolean
+//    fun deleteBook(title: String): Boolean
+//    fun updateBook(book: BookLocal, newBook: BookLocal): Int
+//    fun updateBook(selection: String, selectionArgs: Array<String>, newBook: BookLocal): Int
+//    fun updateImageUrl(res: Resource, fileTitle: String, imgUrl: String): Boolean
+//    fun updateImageUrl(selection: String, selectionArgs: Array<String>, imgUrl: String): Boolean
+//    fun markBookAsPrepared(book: BookLocal): Boolean
```
Comment (ghost) [inline app/src/main/java/vn/ztech/software/projectgutenberg/data/repository/source/local/database/bookreading/BookReadingDbDAOInterface.kt]: mình không cần nữa thì bỏ đi anh

File: app/src/main/res/anim/anim_bottom_up.xml [hunk 17]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<set xmlns:android="http://schemas.android.com/apk/res/android">
+    <translate
+        android:fromYDelta="100%p"
+        android:toYDelta="0%p"
+        android:fillAfter="true"
+        android:duration="300"/>
+</set>
```
Comment (ghost) [inline app/src/main/res/anim/anim_bottom_up.xml]: endline nhá a

File: app/src/main/res/anim/anim_bottom_up_at_top.xml [hunk 18]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<set xmlns:android="http://schemas.android.com/apk/res/android">
+    <translate
+        android:fromYDelta="0%p"
+        android:toYDelta="100%p"
+        android:fillAfter="true"
+        android:duration="300"/>
+</set>
```
Comment (ghost) [inline app/src/main/res/anim/anim_bottom_up_at_top.xml]: .

File: app/src/main/res/anim/anim_top_down.xml [hunk 19]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<set xmlns:android="http://schemas.android.com/apk/res/android">
+    <translate
+        android:fromYDelta="0%p"
+        android:toYDelta="100%p"
+        android:fillAfter="true"
+        android:duration="300"/>
+</set>
```
Comment (ghost) [inline app/src/main/res/anim/anim_top_down.xml]: .

File: app/src/main/res/anim/anim_top_down_at_top.xml [hunk 20]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<set xmlns:android="http://schemas.android.com/apk/res/android">
+    <translate
+        android:fromYDelta="100%p"
+        android:toYDelta="0%p"
+        android:fillAfter="true"
+        android:duration="300"/>
+</set>
```
Comment (ghost) [inline app/src/main/res/anim/anim_top_down_at_top.xml]: .

**Review by hungld-1302 — APPROVED** [review]

**Review by minhnn-1279 — APPROVED** [review]

**Review by hungvv-1301 — APPROVED** [review]

**Pull request merged**